### PR TITLE
feat: Better error reporting when features are disabled

### LINKED ISSUE #2971: Unable to dial bootstrap nodes `InvalidKey`
## Summary

I just upgraded my fork to the current `master`, but am now getting the following error when dialing the boostrap nodes

```
 2022-10-03T18:33:11.985412Z DEBUG iroh_bitswap: dial failure QmbLHAnMoJPWSCR5Zhtx6BHJX9KiKNN6tpvbUcqanj75Nb: Transport([("/dnsaddr/bootstrap.libp2p.io/p2p/QmbLHAnMoJPWSCR5Zhtx6BHJX9KiKNN6tpvbUcqanj75Nb", Other(Custom { kind: Other, error: Other(A(B(Apply(InvalidKey)))) }))])
    at iroh-bitswap/src/lib.rs:313

  2022-10-03T18:33:12.132196Z DEBUG iroh_bitswap: dial failure QmQCU2EcMqAqQPR2i9bChDtGNJchTbq5TbXJJ16u19uLTa: Transport([("/dnsaddr/bootstrap.libp2p.io/p2p/QmQCU2EcMqAqQPR2i9bChDtGNJchTbq5TbXJJ16u19uLTa", Other(Custom { kind: Other, error: Other(A(B(Apply(InvalidKey)))) }))])
    at iroh-bitswap/src/lib.rs:313

  2022-10-03T18:33:12.279092Z DEBUG iroh_bitswap: dial failure QmNnooDu7bfjPFoTZYxMNLWUQJyrVwtbZg5gBMjTezGAJN: Transport([("/dnsaddr/bootstrap.libp2p.io/p2p/QmNnooDu7bfjPFoTZYxMNLWUQJyrVwtbZg5gBMjTezGAJN", Other(Custom { kind: Other, error: Other(A(B(Apply(InvalidKey)))) }))])
    at iroh-bitswap/src/lib.rs:313

  2022-10-03T18:33:12.284856Z DEBUG iroh_bitswap: dial failure QmcZf59bWwK5XFi76CZX8cbJ4BhTzzA3gU1ZjYZcYW3dwt: Transport([("/dnsaddr/bootstrap.libp2p.io/p2p/QmcZf59bWwK5XFi76CZX8cbJ4BhTzzA3gU1ZjYZcYW3dwt", Other(Custom { kind: Other, error: Other(A(B(Apply(InvalidKey)))) }))])
```

My current noise setup looks like this:

```rust
let dh_keys = noise::Keypair::<noise::X25519Spec>::new()
            .into_authentic(keypair)
            .expect("Noise key generation failed");

        noise::NoiseConfig::xx(dh_keys).into_authenticated()
```

### PR DESCRIPTION
**Pull request by thomaseizinger**
## Description

In case support for e.g. RSA keys is disabled at compile-time, we will now print a better error message. For example: 

> Failed to dial Some(PeerId("QmcZf59bWwK5XFi76CZX8cbJ4BhTzzA3gU1ZjYZcYW3dwt")): Failed to negotiate transport protocol(s): [(/dnsaddr/bootstrap.libp2p.io/p2p/QmcZf59bWwK5XFi76CZX8cbJ4BhTzzA3gU1ZjYZcYW3dwt): : Handshake failed: Handshake failed: Invalid public key: Key decoding error: RSA keys are unsupported)]

Fixes #2971.

## Links to any relevant issues

Depends-On: https://github.com/libp2p/rust-libp2p/pull/3081

## Change checklist

- [x] I have performed a self-review of my own code
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] A changelog entry has been made in the appropriate crates

### CODE REVIEW

File: core/src/identity.rs
Comment (mxinden) [inline core/src/identity.rs]:
``` diff
-                Err(DecodingError::new("RSA keys are unsupported"))
+                Err(DecodingError::new("RSA keys are unsupported. Did you enable the `rsa` feature?"))
````

How about mentioning the feature flag here and below?

File: swarm/src/lib.rs
Comment (mxinden) [inline swarm/src/lib.rs:2660]: What is an empty error?

**Review by mxinden — COMMENTED** [review]
As always, great work. Thank you for tackling this so thoroughly. Thanks for breaking it up into smaller meaningful commits.

File: core/src/identity.rs
Comment (mxinden) [reply core/src/identity.rs]: Or maybe "Consider enabling the `rsa` feature.".

File: swarm/src/lib.rs
Comment (thomaseizinger) [reply swarm/src/lib.rs:2660]: An error that does not write anything on `Display::fmt`.

File: core/src/identity.rs
Comment (thomaseizinger) [reply core/src/identity.rs]: Thanks, that gave me an idea for a better wording!

Comment (thomaseizinger) [reply core/src/identity.rs]: See [`c81f730` (#2972)](https://github.com/libp2p/rust-libp2p/pull/2972/commits/c81f730aaceef796cae4ebe09155cb318f4095bb).

**Comment by thomaseizinger** [comment]
@mxinden test-plan doesn't seem to be happy :(

https://github.com/libp2p/rust-libp2p/actions/runs/3186899238/jobs/5197936812

**Comment by mxinden** [comment]
CI is failing with:

```
 error: failed to select a version for `zeroize`.
    ... required by package `der v0.6.0`
    ... which satisfies dependency `der = "^0.6"` of package `pkcs8 v0.9.0`
    ... which satisfies dependency `pkcs8 = "^0.9"` of package `sec1 v0.3.0`
    ... which satisfies dependency `sec1 = "^0.3.0"` of package `libp2p-core v0.37.0 (https://github.com/libp2p/rust-libp2p?rev=fd870f8f4ee1fe54ff605869632ea7e21ecfd3b7#fd870f8f)`
    ... which satisfies git dependency `libp2p-core` of package `libp2p v0.49.0 (https://github.com/libp2p/rust-libp2p?rev=fd870f8f4ee1fe54ff605869632ea7e21ecfd3b7#fd870f8f)`
    ... which satisfies git dependency `libp2pmaster` of package `testplan v0.1.0 (/usr/src/testplan/plan)`
versions that meet the requirements `^1.5` are: 1.5.7, 1.5.6, 1.5.5, 1.5.4, 1.5.3

all possible versions conflict with previously selected packages.

  previously selected package `zeroize v1.3.0`
    ... which satisfies dependency `zeroize = "^1"` (locked to 1.3.0) of package `chacha20 v0.8.2`
    ... which satisfies dependency `chacha20 = "^0.8"` (locked to 0.8.2) of package `chacha20poly1305 v0.9.1`
    ... which satisfies dependency `chacha20poly1305 = "^0.9"` (locked to 0.9.1) of package `snow v0.9.0`
    ... which satisfies dependency `snow = "^0.9.0"` (locked to 0.9.0) of package `libp2p-noise v0.35.0`
    ... which satisfies dependency `libp2p-noise = "^0.35.0"` (locked to 0.35.0) of package `libp2p v0.44.0`
    ... which satisfies dependency `libp2pv0440 = "^0.44.0"` (locked to 0.44.0) of package `testplan v0.1.0 (/usr/src/testplan/plan)`

failed to select a version for `zeroize` which could resolve this conflict
```

I would like to test this with a fresh `Cargo.lock`. I am guessing that previously cached runs influence this one. @laurentsenta is there anyway I can run the CI job without any cache from previous runs?

**Comment by mxinden** [comment]
I should have included this in https://github.com/libp2p/rust-libp2p/pull/2931 but I forgot.

Can you update the pull request now that `v0.49.0` is released @thomaseizinger?

**Comment by thomaseizinger** [comment]
I bumped the version to `0.40.1`. It is my understanding that because `NoiseError` is marked as `non_exhaustive`, this actually is not a breaking change.

**Review by mxinden — APPROVED** [review]
`libp2p-noise` version bump looks good to me.

Either this pull request or some other pull request needs to bump the `libp2p-core` version. Otherwise this change will not be released within the next release round.

Also I don't we should merge here before https://github.com/libp2p/test-plans/issues/59.

Otherwise this looks good to me. Thanks. And again, I wished I merged this before the last release.

**Comment by thomaseizinger** [comment]
> CI is failing with:
> 
> ```
>  error: failed to select a version for `zeroize`.
>     ... required by package `der v0.6.0`
>     ... which satisfies dependency `der = "^0.6"` of package `pkcs8 v0.9.0`
>     ... which satisfies dependency `pkcs8 = "^0.9"` of package `sec1 v0.3.0`
>     ... which satisfies dependency `sec1 = "^0.3.0"` of package `libp2p-core v0.37.0 (https://github.com/libp2p/rust-libp2p?rev=fd870f8f4ee1fe54ff605869632ea7e21ecfd3b7#fd870f8f)`
>     ... which satisfies git dependency `libp2p-core` of package `libp2p v0.49.0 (https://github.com/libp2p/rust-libp2p?rev=fd870f8f4ee1fe54ff605869632ea7e21ecfd3b7#fd870f8f)`
>     ... which satisfies git dependency `libp2pmaster` of package `testplan v0.1.0 (/usr/src/testplan/plan)`
> versions that meet the requirements `^1.5` are: 1.5.7, 1.5.6, 1.5.5, 1.5.4, 1.5.3
> 
> all possible versions conflict with previously selected packages.
> 
>   previously selected package `zeroize v1.3.0`
>     ... which satisfies dependency `zeroize = "^1"` (locked to 1.3.0) of package `chacha20 v0.8.2`
>     ... which satisfies dependency `chacha20 = "^0.8"` (locked to 0.8.2) of package `chacha20poly1305 v0.9.1`
>     ... which satisfies dependency `chacha20poly1305 = "^0.9"` (locked to 0.9.1) of package `snow v0.9.0`
>     ... which satisfies dependency `snow = "^0.9.0"` (locked to 0.9.0) of package `libp2p-noise v0.35.0`
>     ... which satisfies dependency `libp2p-noise = "^0.35.0"` (locked to 0.35.0) of package `libp2p v0.44.0`
>     ... which satisfies dependency `libp2pv0440 = "^0.44.0"` (locked to 0.44.0) of package `testplan v0.1.0 (/usr/src/testplan/plan)`
> 
> failed to select a version for `zeroize` which could resolve this conflict
> ```
> 
> I would like to test this with a fresh `Cargo.lock`. I am guessing that previously cached runs influence this one. @laurentsenta is there anyway I can run the CI job without any cache from previous runs?

@mxinden FYI: https://github.blog/changelog/2022-10-20-manage-caches-in-your-actions-workflows-from-web-interface/

**Comment by thomaseizinger** [comment]
@laurentsenta @mxinden The current test-plan setup for Rust is blocking this PR because of a dependency resolution conflict that can likely be resolved by running `cargo update` but I presume, that is not being done because it would invalidate the docker build cache. Can we find a solution to this please to unblock this? Thank you!

**Comment by mxinden** [comment]
@thomaseizinger https://github.com/libp2p/test-plans/pull/72 is merged. Can you update this pull request?

**Review by melekes — APPROVED** [review]
👍

**Comment by thomaseizinger** [comment]
Here we go!

**Comment by mxinden** [comment]
:tada:

**Pull request merged**
